ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -0,0 +1,267 @@
+          case outcome of
+            Reject reason -> pure (state, Types.Rejected (DomainSpecific reason))
+            Accept mkhandler -> do
+              -- If there was no exception, we are now responsible for closing
+              -- the resource. That's done in a finally after the handler's
+              -- action.
+              rec let cleanup = do
+                        closeResource
+                        modifyMVar_ stateVar $ \state' ->
+                          let !state'' = removeConnection connId state'
+                          in  pure state''
+                  handler <- mkhandler thread
+                  thread <- asyncWithUnmask $ \unmask ->
+                    unmask (action handler) `finally` cleanup
+              let conn = Connection
+                    { connectionThread = thread
+                    , connectionHandle = handle handler
+                    }
+                  !state' = insertConnection connId conn state
+              pure (state', Types.Accepted (Accepted (connectionHandle conn)))
+        -- Caller indicates how to make a new resource. Similar story for the
+        -- above Existing case, except that we create the resource first and
+        -- ensure that it gets closed no matter what.
+        New acquire release -> do
+          -- If acquiring the resource fails, we just re-throw the exception.
+          -- Thus `include`ing a new resource is just like bracketing the
+          -- acquire and release: any exception in acquire will be re-thrown.
+          res <- restore acquire
+          outcome <- restore (withResource Outgoing connId res request)
+                       `onException`
+                       release res
+          case outcome of
+            Reject reason -> do
+              release res
+              pure (state, Types.Rejected (DomainSpecific reason))
+            Accept mkhandler -> do
+              -- Just like for existing connections, the resource will be closed
+              -- when the handler's action finishes.
+              rec let cleanup = do
+                        release res
+                        modifyMVar_ stateVar $ \state' ->
+                          let !state'' = removeConnection connId state'
+                          in  pure state''
+                  handler <- mkhandler thread
+                  thread <- asyncWithUnmask $ \unmask ->
+                    unmask (action handler) `finally` cleanup
+              let conn = Connection
+                    { connectionThread = thread
+                    , connectionHandle = handle handler
+                    }
+                  !state' = insertConnection connId conn state
+              pure (state', Types.Accepted (Accepted (connectionHandle conn)))
+      Just numConn -> case resource of
+        -- Do not call _close; the caller is responsible for that, and knows
+        -- it because we give `Rejected`.
```

Comment:
When we reject it with `Domain reason` we also don't close it.

In that case, and in this case of a duplicate connection, the resource was created by some other system (e.g. an accept loop) and this connections manager has decided not to take ownership of it. It wouldn't be right to close it because that eliminates the possibility of doing something else with it like, for example, trying to include it in another connections manager.

NB: in this case (context of this github comment) the `resource` is the one created by some other system, but there is also a resource in `numConn :: NumberedConnection handle` which represents a connection between the same peers (identifier `connId`).

Reply:
I see, it makes sense.